### PR TITLE
Add multi-platform support for GitHub, GitLab, and SourceHut

### PR DESCRIPTION
**Pull request by ealvar3z**
### Summary

This PR extends vis-plug to support repositories from GitHub, GitLab, and SourceHut (git.sr.ht), providing flexibility for users who host plugins on different platforms.

### Key Changes

- **Multi-platform URL detection**: Added `is_supported_git_url` to recognize GitHub, GitLab, and SourceHut URLs.
- **URL Construction**: Updated `get_full_url` to construct URLs based on detected platform, with GitHub as the default when none specified. Special handling is included for SourceHut shorthand (~user/repo).
- **Repository Name Extraction**: Refined `get_name_from_url` to extract repository names accurately for all platforms.
- **URL Shortening**: Adjusted `get_short_url` to format display paths for each platform.

### Additional Enhancements

Basic error handling was added to `get_full_url` to warn users when an unrecognized URL format is encountered.

This update allows vis-plug users to seamlessly work with plugins hosted across multiple git platforms.

### CODE REVIEW

File: init.lua [hunk 1]
```diff
@@ -88,8 +93,32 @@ local get_file_path = function(plug)
...
+-- '{http[s]://git.sr.ht}/~emg/vis-cscope.git'
+local is_supported_git_url = function(url)
 	return url:find('^http[s]?://github.com.+') ~= nil
+    	or url:find('^http[s]?://gitlab.com.+') ~= nil
+    	or url:find('^http[s]?://git.sr.ht.+') ~= nil
+end
+
+-- get the full url for supported platforms
+local get_full_url = function(url)
+    if is_supported_git_url(url) then
```
Comment (erf) [inline init.lua]: Should we limit ourselves to these 3 or should we allow just any URL ?

File: init.lua [hunk 2]
```diff
@@ -88,8 +93,32 @@ local get_file_path = function(plug)
...
+      return url
+    elseif is_no_host_url(url) then
+      if url:find('^~') then
```
Comment (erf) [inline init.lua]: Should we move this to a separate method? Should we also mention the supported short URL's and their format in the README.md ?

File: init.lua (hunk 1 above)
Comment (erf) [reply init.lua]: Perhaps a single regex could be used for these 3 ? Something like:
```
return url:find('^http[s]?://(github%.com|gitlab%.com|git%.sr%.ht).+') ~= nil
```

**Review by erf — COMMENTED** [review]
Thanks for the contribution.

I made a few comments in the code. 

Maybe you also could describe in the PR a few examples of how to support the new supported URL's and maybe also add some examples to `README.md`.

File: init.lua (hunk 2 above)
Comment (erf) [reply init.lua]: Something like: `get_full_url_from_no_host_url`

File: init.lua [hunk 3]
```diff
@@ -88,8 +93,32 @@ local get_file_path = function(plug)
...
+        -- takes care of sourcehuts ~USER fmt
+        return 'https://git.sr.ht/' .. url
+      else
+        -- default to github
+        return 'https://github.com/' .. url
```
Comment (erf) [inline init.lua]: Hmm.. can this be right? This used to be:
```
	elseif is_no_host_url(url) then
		return 'https://' .. url
```

**Comment by ealvar3z** [comment]
> Thanks for the contribution.
> 
> I made a few comments in the code.
> 
> Maybe you also could describe in the PR a few examples of how to support the new supported URL's and maybe also add some examples to `README.md`.

Will do, thank you for the review. Originally, I was just trying to get an MVP to get myself going and then decided to just PR the change.

File: init.lua (hunk 3 above)
Comment (ealvar3z) [reply init.lua]: hmm.. i see, you're correct. therefore, i'll revert to the original handling of `is_no_host_url` and use the existing platform defaults only when appropriate. Here’s what i have in mind for `get_full_url`:

```lua
local get_full_url = function(url)
    if is_supported_git_url(url) then
        return url
    elseif is_no_host_url(url) then
        -- For SourceHut shorthand (~user) or any unrecognized host, just prepend 'https://'
        if url:find('^~') then
            return 'https://git.sr.ht/' .. url
        else
            return 'https://' .. url  -- Keeps the original host structure
        end
    elseif is_short_ssh_url(url) then
        return url
    else
        -- Log a warning for unexpected URL formats and default to GitHub
        vis:message("Warning: Unrecognized URL format, defaulting to GitHub.")
        return 'https://github.com/' .. url
    end
end
```

**Comment by erf** [comment]
> > Thanks for the contribution.
> 
> > 
> 
> > I made a few comments in the code.
> 
> > 
> 
> > Maybe you also could describe in the PR a few examples of how to support the new supported URL's and maybe also add some examples to `README.md`.
> 
> 
> 
> Will do, thank you for the review. Originally, I was just trying to get an MVP to get myself going and then decided to just PR the change.

Nice. I haven't really tested this myself yet. Just had a brief look at the changes. 

Btw. Ideally we should add regression tests to avoid causing issues on changes.

File: init.lua [hunk 4]
```diff
@@ -12,22 +12,6 @@ local plugins_path = nil
 -- we store commands in an array of tables {name, func, desc}
 local commands = {}
 
--- add a searcher to find plugins by filename, not module path
-table.insert(package.searchers, function(name)
```
Comment (erf) [inline init.lua:16]: why these changes?

File: init.lua [hunk 5]
```diff
@@ -37,9 +21,26 @@ local get_cache_path = function()
 	return CACHE_DIR .. '/vis-plug'
 end
 
+-- set custom path and add it first to package.path for require
+M.path = function(path)
+	plugins_path = path or get_cache_path()
```
Comment (erf) [inline init.lua:26]: what is going on here? this looks like a separate change that needs explanation

File: init.lua [hunk 6]
```diff
@@ -206,15 +193,20 @@ local checkout = function(plug)
...
 local plug_require = function(plug, args)
 	if not file_exists(plug.path) then
+    vis:message("Warning: Plugin path not found: " .. plug_path)
 		return
 	end
 	if plug.theme then
 		return
 	end
-	local name = 'plugins/' .. plug.name .. plug.file
+	local name = file_exists(plug.path .. '/' .. plug.name .. '.lua')
```
Comment (erf) [inline init.lua:205]: don't know if this is related to said change

File: init.lua [hunk 7]
```diff
@@ -87,9 +93,19 @@ local get_file_path = function(plug)
...
 
--- '{http[s]://github.com}/erf/vis-cursors.git'
-local is_github_url = function(url)
-	return url:find('^http[s]?://github.com.+') ~= nil
+-- check if url is fully qualified
+local is_fully_qualified = function(url)
+  return url:find('^https[s]?://') ~= nil
+end
+
+local get_full_url_from_no_host_url = function(url)
```
Comment (erf) [inline init.lua:101]: this is not called by anyone ..?

**Comment by erf** [comment]
There is too much stuff going on in this PR - closing

If you want to contribute, please make a focused PR / commits. Until it's ready, mark it as **Draft**.

**Comment by ealvar3z** [comment]
> There is too much stuff going on in this PR - closing
> 
> If you want to contribute, please make a focused PR / commits. Until it's ready, mark it as **Draft**.

I understand, the changes were made to ensure that they were no regressions to the original API. if you take a look at the diff, most of it was moving stuff around and ensuring that the paths of the `.cache/plugins` directories would be supported for plugins that do not follow the `<plugin directory>/init.lua` pattern (eg. timoha's vis-go.lua).

As always, thank you for the project, and thank you for the review. I'll maintain my current fork. If you need anything in the future for this project. hit me up.

**Comment by ealvar3z** [comment]
@erf , so i did some digging and after doing a vanilla install of both `vis` and use `vis-plug` as you originally wrote it. It already supports everything out of the box. Hindsight being 20/20, what I should've done is exhaust the plugin like I just did and then do a PR for the README. Thereby,  exhausting all of the scenarios that I _implemented_ in my defunct PR.

As always, thank you for your patience, review and the `vis-plug` project. I will delete the branch on my end and use the plugin as originally intended.

**Comment by erf** [comment]
I see . Well good things just work. I could perhaps look into shortcuts for some other git sites (this should not work per now), but for now i mostly use github and full urls should work as well.
